_posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown

```diff
@@ -0,0 +1,160 @@
+
+In this blog post, we'll introduce the concept of extensions and outline some proposed projects in this area.
+
+## Introducing extensions
+
+From the product point of view, extensions are a new mechanism that provides a way to break up a monolithic, tightly coupled model of building new features in OpenSearch. Technically, extensions are a simple evolution of plugins --- or plugins decoupled from their OpenSearch/Dashboards hosts. In practice, an extension is only different from a plugin in that it only depends on the OpenSearch/Dashboards SDK, and works with multiple versions of OpenSearch/Dashboards. We aim for the existence of many more extensions than there are plugins today, written by many more developers. Then we will sunset plugins.
+
+We want for extensions to become the preferred mechanism for providing functionality in OpenSearch/Dashboards. Having multiple competing implementations will produce extensions with high performance, improved security, and versatile features for all categories of users, administrators and developers.
+
+## Extensions roadmap
+
+The following proposed projects are in chronological order, but many of them can be accomplished in parallel.
+
+### Provide experimental SDKs and move extensibility concerns out of cores
+
+First, we plan to introduce an OpenSearch SDK and an OpenSearch Dashboards SDK, and refactor the OpenSearch/Dashboards cores to support them as needed. This creates both a logical and a physical separation between extensions and their hosts. You should be able to author an extension that is compatible with all minor versions of an OpenSearch/Dashboards release and to upgrade OpenSearch/Dashboards without having to upgrade an installed extension.
+
+The SDK assumes current and future extensibility concerns from OpenSearch/Dashboards. It will contain the set of APIs that need to follow semver, significantly reducing the number of APIs that OpenSearch/Dashboards needs to worry about, because plugins only take a dependency on the SDK. With a semver-stable SDK, plugins can confidently declare that they work with, for example, `OpenSearch/Dashboards >= 2.3.0` (2.3, 2.4, ... 3.0, and so on) or `~> 2.5` (any 2.x after 2.5). The SDK will provide support for integration testing against broad ranges of OpenSearch/Dashboards versions. It can begin selecting common functionality that all plugins may need, such as storing credentials or saving objects, and be strongly opinionated about what constitutes a semver-compatible extension point for OpenSearch/Dashboards. Instead of importing a transitive dependency (e.g. `oui`), developers import an SDK namespace (e.g. `sdk/ui`).
```

Comment:
```suggestion
The SDK assumes current and future extensibility concerns from OpenSearch/Dashboards. It will contain the set of APIs that need to follow semver, significantly reducing the number of APIs that OpenSearch/Dashboards needs to worry about, because plugins only take a dependency on the SDK. With a semver-stable SDK, plugins can confidently declare that they work with, for example, `OpenSearch/Dashboards >= 2.3.0` (2.3, 2.4, ... 3.0, and so on) or `~> 2.5` (any 2.x after 2.5). The SDK will provide support for integration testing against broad ranges of OpenSearch/Dashboards versions. It can begin selecting common functionality that all plugins may need, such as storing credentials or saving objects, and be strongly opinionated about what constitutes a semver-compatible extension point for OpenSearch/Dashboards. Instead of importing a transitive dependency (for example, `oui`), developers import an SDK namespace (for example, `sdk/ui`).
```